drivers/usb/gadget/function/f_scm.c

```diff
@@ -589,7 +647,417 @@ static void __exit f_scm_exit(void)
 module_init(f_scm_init);
 module_exit(f_scm_exit);
 
+/* Handle USB listening and writing */
+static void scm_send_msg_complete(struct usb_ep *ep, struct usb_request *req)
+{
+}
+static void scm_send_msg(void *data, int len, struct f_scm *scm_inst)
+{
+	struct usb_request *req = scm_inst->req_in;
+	int ret;
+
+	if (!req)
+		return;
+
+	req->buf = kmalloc(MAX_INT_PACKET_SIZE, GFP_ATOMIC);
+	memcpy(req->buf, data, len);
+	req->length = len;
+	ret = usb_ep_queue(scm_inst->cmd_in, scm_inst->req_in, GFP_ATOMIC);
+}
+
+/* Reads SCM command from the host */
+static void scm_process_out_cmd(struct scm_packet *packet, size_t len,
+	struct f_scm *usb_context)
+{
+	/**
+	 *Make sure the packet is big enough for the packet and payload
+	 * (checked in order to avoid reading bad memory)
+	 */
+	if (!packet || len < sizeof(*packet) ||
+		len < (sizeof(*packet)+packet->hdr.payload_len))
+		return;
+
+	/* Incoming command is either a close notificaiton or ACK */
+	switch (packet->hdr.opcode) {
+	case SCM_OP_ACK:
+		scm_proxy_recv_ack(packet, usb_context->proxy_context);
+		break;
+	case SCM_OP_CLOSE:
+	default:
+		break;
+	}
+}
+static void scm_read_out_cmd_cb(struct usb_ep *ep, struct usb_request *req)
+{
+	if (req->buf)
+		scm_process_out_cmd(req->buf, req->actual, req->context);
+	scm_read_out_cmd(req->context);
+}
+static int scm_read_out_cmd(struct f_scm *scm_inst)
+{
+	struct usb_request *out_req = scm_inst->req_out;
+
+	out_req->length = sizeof(struct scm_packet) + 64;
+	out_req->buf = kmalloc(out_req->length, GFP_ATOMIC);
+	out_req->dma = 0;
+	out_req->complete = scm_read_out_cmd_cb;
+	out_req->context = scm_inst;
+	usb_ep_queue(scm_inst->cmd_out, out_req, GFP_ATOMIC);
+
+	return 0;
+}
+
+
+/* SCM Proxy */
+/* This definately has to be moved somewhere else */
 MODULE_LICENSE("GPL v2");
 MODULE_AUTHOR("Daniel Berliner");
 MODULE_DESCRIPTION("SCM Driver");
 MODULE_VERSION("0.0.1");
+
+
+/* SCM Proxy internal functions */
+struct scm_proxy_inst {
+	void *usb_context;
+	__u8 scm_msg_id;
+	struct mutex scm_msg_id_mutex;
```

Comment:
This is a perfect use case for an atomic counter, instead of the more heavy wait mutex, I think.
 
Replace `__u8 scm_msg_id` with `atomic_t` and use `atomic_inc_ret` to increment and fetch the value. 

If that atomic is too hard deal with, at least replace the mutex with a spinlock.  No need to support blocking/sleeping the task when the lock is only held for one instruction.

Reply:
Agreed, that is the best thing to use here. I also changed the value to 16 bit to match the updated spec.